File: tests/test_check_prod_test_separation.py
#!/usr/bin/env python3
import importlib.util
import json
import sys
import tempfile
from pathlib import Path
import unittest


def _load_module():
    repo_root = Path(__file__).resolve().parents[1]
    module_path = repo_root / "scripts" / "check_prod_test_separation.py"
    spec = importlib.util.spec_from_file_location(
        "check_prod_test_separation", module_path
    )
    if spec is None or spec.loader is None:
        raise RuntimeError(f"Unable to load module from {module_path}")
    module = importlib.util.module_from_spec(spec)
    sys.modules[spec.name] = module
    spec.loader.exec_module(module)
    return module


checker = _load_module()


class ProdTestSeparationCheckTests(unittest.TestCase):
    def test_test_like_filenames_are_detected(self):
        self.assertTrue(checker._is_test_like_filename("tests.rs"))
        self.assertTrue(checker._is_test_like_filename("foo_tests.rs"))
        self.assertTrue(checker._is_test_like_filename("test_helper.rs"))
        self.assertFalse(checker._is_test_like_filename("module.rs"))

    def test_detects_test_file_and_symbol_reexport(self):
        with tempfile.TemporaryDirectory(prefix="zkfuzz_prod_test_sep_") as tmpdir:
            root = Path(tmpdir)
            (root / "src").mkdir(parents=True, exist_ok=True)
            (root / "src" / "mod.rs").write_text(
                "pub(super) use super::attack_runner_tests::helper;\n",
                encoding="utf-8",
            )
            (root / "src" / "attack_runner_tests.rs").write_text(
                "pub fn helper() {}\n", encoding="utf-8"
            )

            violations = checker.collect_violations(root, ["src"])
            kinds = {v.kind for v in violations}
            self.assertIn("test_symbol_import_or_reexport", kinds)
            self.assertIn("test_file_in_production_tree", kinds)

    def test_cfg_test_is_not_allowed_in_production(self):
        with tempfile.TemporaryDirectory(prefix="zkfuzz_prod_test_sep_") as tmpdir:
            root = Path(tmpdir)
            (root / "src").mkdir(parents=True, exist_ok=True)
            (root / "src" / "mod.rs").write_text(
                '#[cfg(test)]\n#[path = "mod_tests.rs"]\nmod tests;\npub fn f() {}\n',
                encoding="utf-8",
            )

            violations = checker.collect_violations(root, ["src"])
            kinds = {v.kind for v in violations}
            self.assertIn("test_attribute_in_production", kinds)
            self.assertIn("test_path_attr_in_production", kinds)
            self.assertIn("test_module_decl_in_production", kinds)

    def test_detects_test_module_without_cfg(self):
        with tempfile.TemporaryDirectory(prefix="zkfuzz_prod_test_sep_") as tmpdir:
            root = Path(tmpdir)
            (root / "src").mkdir(parents=True, exist_ok=True)
            (root / "src" / "mod.rs").write_text(
                '#[path = "mod_tests.rs"]\nmod tests;\n',
                encoding="utf-8",
            )

            violations = checker.collect_violations(root, ["src"])
            kinds = [v.kind for v in violations]
            self.assertIn("test_module_decl_in_production", kinds)
            self.assertIn("test_path_attr_in_production", kinds)

    def test_main_passes_on_clean_source(self):
        with tempfile.TemporaryDirectory(prefix="zkfuzz_prod_test_sep_") as tmpdir:
            root = Path(tmpdir)
            (root / "src").mkdir(parents=True, exist_ok=True)
            (root / "src" / "mod.rs").write_text("pub fn f() {}\n", encoding="utf-8")

            exit_code = checker.main(["--repo-root", str(root), "--search-roots", "src"])
            self.assertEqual(exit_code, 0)

    def test_main_uses_baseline_and_rejects_new_violation(self):
        with tempfile.TemporaryDirectory(prefix="zkfuzz_prod_test_sep_") as tmpdir:
            root = Path(tmpdir)
            (root / "src").mkdir(parents=True, exist_ok=True)
            baseline_path = root / "config" / "prod_test_separation_baseline.json"
            baseline_path.parent.mkdir(parents=True, exist_ok=True)

            (root / "src" / "legacy.rs").write_text(
                "#[cfg(test)]\npub fn legacy_only() {}\n", encoding="utf-8"
            )
            checker.main(
                [
                    "--repo-root",
                    str(root),
                    "--search-roots",
                    "src",
                    "--baseline",
                    str(baseline_path),
                    "--write-baseline",
                ]
            )

            # Baseline covers current state.
            exit_code = checker.main(
                [
                    "--repo-root",
                    str(root),
                    "--search-roots",
                    "src",
                    "--baseline",
                    str(baseline_path),
                ]
            )
            self.assertEqual(exit_code, 0)

            # Increase count for an existing signature (same file/kind/code).
            (root / "src" / "legacy.rs").write_text(
                "#[cfg(test)]\npub fn legacy_only() {}\n#[cfg(test)]\npub fn legacy_two() {}\n",
                encoding="utf-8",
            )
            exit_code = checker.main(
                [
                    "--repo-root",
                    str(root),
                    "--search-roots",
                    "src",
                    "--baseline",
                    str(baseline_path),
                ]
            )
            self.assertEqual(exit_code, 1)

            # Restore to baseline state before testing brand-new signature.
            (root / "src" / "legacy.rs").write_text(
                "#[cfg(test)]\npub fn legacy_only() {}\n", encoding="utf-8"
            )

            # Introduce new violation not present in baseline.
            (root / "src" / "new_bad.rs").write_text(
                "#[cfg(test)]\npub fn new_bad() {}\n", encoding="utf-8"
            )
            exit_code = checker.main(
                [
                    "--repo-root",
                    str(root),
                    "--search-roots",
                    "src",
                    "--baseline",
                    str(baseline_path),
                ]
            )
            self.assertEqual(exit_code, 1)

            baseline = json.loads(baseline_path.read_text(encoding="utf-8"))
            self.assertTrue(baseline["violations"])


if __name__ == "__main__":
    unittest.main()

File: tests/test_check_prod_test_separation.rs
use tempfile::tempdir;
use zk_fuzzer::checks::prod_test_separation::{
    collect_violations, filter_new_violations, is_test_like_filename, load_baseline, write_baseline,
};

#[test]
fn test_like_filenames_are_detected() {
    assert!(is_test_like_filename("tests.rs"));
    assert!(is_test_like_filename("foo_tests.rs"));
    assert!(is_test_like_filename("test_helper.rs"));
    assert!(!is_test_like_filename("module.rs"));
}

#[test]
fn detects_test_file_and_symbol_reexport() {
    let tmp = tempdir().expect("tempdir");
    let root = tmp.path();
    std::fs::create_dir_all(root.join("src")).expect("mkdir src");
    std::fs::write(
        root.join("src/mod.rs"),
        "pub(super) use super::attack_runner_tests::helper;\n",
    )
    .expect("write mod");
    std::fs::write(
        root.join("src/attack_runner_tests.rs"),
        "pub fn helper() {}\n",
    )
    .expect("write tests module");

    let violations = collect_violations(root, &["src".to_string()]).expect("collect violations");
    let kinds: std::collections::BTreeSet<&str> =
        violations.iter().map(|v| v.kind.as_str()).collect();
    assert!(kinds.contains("test_symbol_import_or_reexport"));
    assert!(kinds.contains("test_file_in_production_tree"));
}

#[test]
fn detects_cfg_test_and_test_path_attr_and_module() {
    let tmp = tempdir().expect("tempdir");
    let root = tmp.path();
    std::fs::create_dir_all(root.join("src")).expect("mkdir src");
    std::fs::write(
        root.join("src/mod.rs"),
        "#[cfg(test)]\n#[path = \"mod_tests.rs\"]\nmod tests;\npub fn f() {}\n",
    )
    .expect("write mod");

    let violations = collect_violations(root, &["src".to_string()]).expect("collect violations");
    let kinds: std::collections::BTreeSet<&str> =
        violations.iter().map(|v| v.kind.as_str()).collect();
    assert!(kinds.contains("test_attribute_in_production"));
    assert!(kinds.contains("test_path_attr_in_production"));
    assert!(kinds.contains("test_module_decl_in_production"));
}

#[test]
fn baseline_flow_detects_new_violations() {
    let tmp = tempdir().expect("tempdir");
    let root = tmp.path();
    std::fs::create_dir_all(root.join("src")).expect("mkdir src");
    std::fs::create_dir_all(root.join("config")).expect("mkdir config");
    let baseline_path = root.join("config/prod_test_separation_baseline.json");

    std::fs::write(
        root.join("src/legacy.rs"),
        "#[cfg(test)]\npub fn legacy_only() {}\n",
    )
    .expect("write legacy");
    let initial = collect_violations(root, &["src".to_string()]).expect("collect initial");
    write_baseline(&baseline_path, &initial).expect("write baseline");

    let baseline_counts = load_baseline(&baseline_path).expect("load baseline");
    let new_from_initial = filter_new_violations(&initial, &baseline_counts);
    assert!(new_from_initial.is_empty());

    std::fs::write(
        root.join("src/new_bad.rs"),
        "#[cfg(test)]\npub fn new_bad() {}\n",
    )
    .expect("write new bad");
    let with_new = collect_violations(root, &["src".to_string()]).expect("collect with new");
    let new_violations = filter_new_violations(&with_new, &baseline_counts);
    assert!(!new_violations.is_empty());
}
